lib/components/mixins/tapable.dart

```diff
@@ -1,74 +1,106 @@
-import 'dart:ui';
-
 import 'package:flutter/foundation.dart';
 import 'package:flutter/gestures.dart';
 
+import '../base_component.dart';
+import '../component.dart';
+import '../../extensions/offset.dart';
 import '../../game/base_game.dart';
-import '../position_component.dart';
 
-mixin Tapable on PositionComponent {
-  void onTapCancel() {}
-  void onTapDown(TapDownDetails details) {}
-  void onTapUp(TapUpDetails details) {}
+mixin Tapable on BaseComponent {
+  bool onTapCancel() {
+    return true;
+  }
+
+  bool onTapDown(TapDownDetails details) {
+    return true;
+  }
+
+  bool onTapUp(TapUpDetails details) {
+    return true;
+  }
 
   int _currentPointerId;
 
   bool _checkPointerId(int pointerId) => _currentPointerId == pointerId;
 
-  bool checkTapOverlap(Rect rect, Offset o) => rect.contains(o);
-
-  void handleTapDown(Rect rect, int pointerId, TapDownDetails details) {
-    if (checkTapOverlap(rect, details.localPosition)) {
+  bool handleTapDown(int pointerId, TapDownDetails details) {
+    if (checkOverlap(details.localPosition.toVector2())) {
       _currentPointerId = pointerId;
-      onTapDown(details);
+      return onTapDown(details);
     }
+    return true;
   }
 
-  void handleTapUp(Rect rect, int pointerId, TapUpDetails details) {
+  bool handleTapUp(int pointerId, TapUpDetails details) {
     if (_checkPointerId(pointerId) &&
-        checkTapOverlap(rect, details.localPosition)) {
+        checkOverlap(details.localPosition.toVector2())) {
       _currentPointerId = null;
-      onTapUp(details);
+      return onTapUp(details);
     }
+    return true;
   }
 
-  void handleTapCancel(int pointerId) {
+  bool handleTapCancel(int pointerId) {
     if (_checkPointerId(pointerId)) {
       _currentPointerId = null;
-      onTapCancel();
+      return onTapCancel();
     }
+    return true;
   }
 }
 
 mixin HasTapableComponents on BaseGame {
-  Iterable<PositionComponent> _positionComponents() {
-    return components.whereType<PositionComponent>();
-  }
-
   @mustCallSuper
   void onTapCancel(int pointerId) {
-    _positionComponents().forEach((c) {
-      c.propagateToChildren<Tapable>(
-        (child, _) => child.handleTapCancel(pointerId),
-      );
-    });
+    for (Component c in components.toList().reversed) {
+      bool shouldContinue = true;
+      if (c is BaseComponent) {
+        shouldContinue = c.propagateToChildren<Tapable>(
+          (child) => child.handleTapCancel(pointerId),
+        );
```

Comment:
you can just call if(xxx) break here directly and also on the other one and don't need the variable

Reply:
I had it like that, but this was added to add readability of the code.

Reply:
they are both readable to me, but no strong opinions